### ISSUE TITLE
How to include background images

### ISSUE THREAD

**Issue by samuelli66** [pierluigiferrari/ssd_keras#117]
First, thank you so much for this fantastic work!!

Here is my question. I need to include some background images to the training data generator, because they contain things that can be easily confused with the target objects. Consider them as the hard negative samples. During the hard negative mining, SSD should be able to pick them up and decreases the false positives. I am currently using the parse_csv tool. In the label.csv, they look like the following:

image_name, class_id, xmin, ymin, xmax, ymax
image1.jpg, 0, 0, 0, 0, 0

And I set `include_class = [0, 1, 2, 3]` in the data generator.

Is this the right way? Could you please suggest other ways to include those images in the data generator? 

Thanks!

**Comment by pierluigiferrari** [comment]
Out of the three parsers, the CSV parser is the only one that doesn't take the possibility of images without any ground truth into account.

However, with a few workarounds it'll still work. It should actually work exactly the way you're doing it, but you need to make sure that a few parameters are set correctly.

The way you're doing it, after parsing the annotations, the list of labels for each background image will contain one box of class 0 with box coordinates (0, 0, 0, 0), i.e. a degenerate bounding box. This works out fine, because `DataGenerator` can automatically remove degenerate boxes from the ground truth, leaving the background images with no ground truth at all, which is exactly what you want. With the following settings, everything should work out:

1. In the data generator's `generate()` method, make sure `degenerate_box_handling` is set to 'remove'. This is the default value anyway, but it's important. As explained above, this makes sure that the (0, 0, 0, 0) dummy boxes are being removed.
2. Also in the data generator's `generate()` method, set `keep_images_without_gt=True`. As the argument name says, this makes sure that the generator keeps an image in the batch even if there is no ground truth for it.
3. In the constructor of `SSDLoss`, set `n_neg_min` to some reasonable number greater than zero. This could be important, because in its default settings, the loss function will only take at most 3 times as many negative boxes into account as there are positive boxes. This means in case there are 0 positive boxes in a batch, then there are also 0 negative boxes going into the loss. `n_neg_min` makes sure that a certain minimum number of negative boxes always goes into the loss. You could set `n_neg_min=batch_size` for example.

**Comment by samuelli66** [comment]
This is great. Thanks for your prompt and detailed explanation!
